hal/armv7m/mpu.c

```diff
@@ -0,0 +1,333 @@
+/*
+ * Phoenix-RTOS
+ *
+ * Phoenix-RTOS loader
+ *
+ * MPU routines
+ *
+ * Copyright 2021 Phoenix Systems
+ * Author: Gerard Swiderski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include <lib/lib.h>
+#include <lib/log.h>
+#include <lib/errno.h>
+#include <hal/hal.h>
+#include <syspage.h>
+#include "mpu.h"
+
+enum {
+	mpu_fError = 0x01,
+	mpu_fConfigured = 0x04,
+	mpu_fValid = 0x08,
+};
+
+
+struct {
+	syspage_hal_t hal;
+	u8 mapsCnt;
+	u8 regCnt;
+	u8 regMax;
+	u8 flags;
+} mpu_common;
+
+
+/* clang-format off */
+
+/* Binary count trailing zeros */
+__attribute__((always_inline)) static inline u8 _ctz(u32 val)
+{
+	asm volatile(" \
+		rbit %0, %0; \
+		clz %0, %0"
+		: "+l"(val));
+	return val;
+}
```

Comment:
This could be added to hal

Reply:
@HBuczynski what do You think, it is local to `mpu.c`, but may be reused later (it's like `__bultin_ctz()`, maybe `__builtin`is better?)

Reply:
> This could be added to hal

What do you mean? Would you like to add this function to `hal.h`?
It is only used in MPU in hal/armv7m so I think that this function is in a good place.

Reply:
Guys please don't add function specific for one architecture as function with generalized interface. If you need specific function when you implement specific hardware functionality you can add it, but nothing more.

Reply:
@pawelpisarczyk such function is a part of hal in phoenix-rtos-kernel (https://github.com/phoenix-rtos/phoenix-rtos-kernel/blob/964cd68eac3da0175fd98abc25cb65538dbf1825/hal/ia32/cpu.h#L422). Why it can't be in plo?

Reply:
@agkaminski In kernel function `hal_cpuGetLastBit()` is used by hardware independent part. This means that it should be implemented in HAL layer of each supported architecture. In this case this function is used only in this particular HAL by the hardware dependent module handling MPU and it is called two times. This is the main and important difference. When this function will be used by hardware independent part of plo by any function then it can be treated as the part of generalized HAL interface. Currently I see no reason to extend the HAL interface with it.

Reply:
Ok, now it's clear to me. It is important though to remember that and should this function be needed in other parts of code move it to HAL.